test/georef_ocd_mapping_t.cpp

```diff
@@ -0,0 +1,318 @@
+	        << GeorefCoreData { 15000, 5.22, 5.22, { 52.39963782, 13.04811373 }, QStringLiteral("UTM"), { QStringLiteral("33") } };
+	QTest::newRow("Nasca airport - Peru, WGS 84/UTM zone 18S")
+	        << QStringLiteral("\tm15000\tg50.0000\tr1\tx504606\ty8358028\ta-2.70000000\td1000.000000\ti-2018\tb0.00\tc0.00")
+	        << 11006
+	        << GeorefCoreData { 15000, -2.70, -2.70, { -14.85213486, -74.95718575 }, QStringLiteral("UTM"), { QStringLiteral("18 S") } };
+	QTest::newRow("Wellington airport - New Zealand, WGS 84/UTM zone 60S")
+	        << QStringLiteral("\tm15000\tg50.0000\tr1\tx316493\ty5423682\ta21.14000000\td1000.000000\ti-2060\tb0.00\tc0.00")
+	        << 11006
+	        << GeorefCoreData { 15000, 21.14, 21.14, { -41.31746197, 174.80754616 }, QStringLiteral("UTM"), { QStringLiteral("60 S") } };
+}
+
+void GeoreferenceMappingTest::testOcdToMapper_data()
+{
+	commonTestData();
+
+	QTest::newRow("OCD CRS Local")
+	        << QStringLiteral("\tm15000\tg50.0000\tr1\tx5556666\ty6667777\ta15.35000000\td1000.000000\ti1000\tb12.00\tc16.00")
+	        << 11006
+	        << GeorefCoreData { 15000, 15.35, 15.35, { 0, 0 }, QStringLiteral("Local"), { } };
+	// exercise error paths
+	QTest::newRow("Unknown OCD grid/zone combination")
+	        << QStringLiteral("\tm15000\tg50.0000\tr1\tx5556666\ty6667777\ta15.35000000\td1000.000000\ti99999\tb12.00\tc16.00")
+	        << 11006
+	        << GeorefCoreData { 15000, 15.35, 15.35, { 0, 0 }, QStringLiteral("Local"), { } };
+}
+
+// Extract data from OCD string type 1039
+// Must return unitialized structure for empty string like the translation functions do for invalid input.
+static auto fieldsFromString(const QString& scalepar_string)
+{
+	OcdGeoref::OcdStringFields fields;
+
+	for (const auto& param : scalepar_string.splitRef(QLatin1Char('\t'), QString::SkipEmptyParts))
+	{
+		int* int_param_ptr = nullptr;
+		bool ok;
+
+		switch (param.at(0).unicode())
+		{
+		case 'm':
+			int_param_ptr = &fields.m;
+			// fallthrough
+		case 'x':
+			if (!int_param_ptr)
+				int_param_ptr = &fields.x;
+			// fallthrough
+		case 'y':
+			if (!int_param_ptr)
+				int_param_ptr = &fields.y;
+			// fallthrough
+		case 'i':
+			if (!int_param_ptr)
+				int_param_ptr = &fields.i;
+			{
+				auto int_param = param.mid(1).toInt(&ok);
+				if (ok)
+					*int_param_ptr = int_param;
+				int_param_ptr = nullptr;
+			}
+			break;
+		case 'a':
+			{
+				auto double_param = param.mid(1).toDouble(&ok);
+				if (ok)
+					fields.a = double_param;
+			}
+			break;
+		default:
+			; // ignore unknown and unhandled fields
+		}
+	}
+
+	return fields;
+}
+
+Q_DECLARE_METATYPE(OcdGeoref::OcdStringFields)
+
+bool operator==(const OcdGeoref::OcdStringFields& lhs, const OcdGeoref::OcdStringFields& rhs)
+{
+	return lhs.i == rhs.i
+	        && lhs.m == rhs.m
+	        && lhs.x == rhs.x
+	        && lhs.y == rhs.y
+	        && ((qIsNaN(lhs.a) && qIsNaN(rhs.a)) || qAbs(lhs.a - rhs.a) < 5e-9); // 8-digit precision or both NaN's
+}
```

Comment:
How about moving this function to ocd_georef.h/.cpp? It contains assumptions about OcdGeoref...Fields (precision) which belong there. And it would stop someone else from adding it there.

Reply:
Makes a lot of sense to me. The operator is generic to the type.